src/Hosting/TestHost/test/TestClientTests.cs

```diff
@@ -162,6 +167,296 @@ public void Dispose()
+
+            await responseStartedSyncPoint.WaitForSyncPoint().WithTimeout();
+            responseStartedSyncPoint.Continue();
+
+            var responseContent = await response.Content.ReadAsStreamAsync().WithTimeout();
+
+            // Assert
+
+            // Ensure request stream has started
+            await requestStreamSyncPoint.WaitForSyncPoint();
+
+            // Write to request
+            await requestStream.WriteAsync(Encoding.UTF8.GetBytes("SENT")).AsTask().WithTimeout();
+            await requestStream.FlushAsync().WithTimeout();
+            await responseReadSyncPoint.WaitForSyncPoint().WithTimeout();
+
+            // Cancel request. Disposing response must be used because SendAsync has finished.
+            response.Dispose();
+            responseReadSyncPoint.Continue();
+
+            await responseEndingSyncPoint.WaitForSyncPoint().WithTimeout();
+            responseEndingSyncPoint.Continue();
+
+            Assert.True(readCanceled);
+
+            requestStreamSyncPoint.Continue();
+        }
+
+        [Fact]
+        public async Task ClientStreaming_ResponseCompletesWithoutReadingRequest()
+        {
+            // Arrange
+            var requestStreamSyncPoint = new SyncPoint();
+
+            RequestDelegate appDelegate = async ctx =>
+            {
+                await ctx.Response.WriteAsync("POST Response");
+            };
+
+            Stream requestStream = null;
+
+            var builder = new WebHostBuilder().Configure(app => app.Run(appDelegate));
+            var server = new TestServer(builder);
+            var client = server.CreateClient();
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, "http://localhost:12345");
+            httpRequest.Version = new Version(2, 0);
+            httpRequest.Content = new PushContent(async stream =>
+            {
+                // Initial flush to ensure headers are sent
+                await stream.FlushAsync();
+
+                requestStream = stream;
+                await requestStreamSyncPoint.WaitToContinue();
+            });
+
+            // Act
+            var response = await client.SendAsync(httpRequest, HttpCompletionOption.ResponseHeadersRead).WithTimeout();
+
+            var responseContent = await response.Content.ReadAsStreamAsync().WithTimeout();
+
+            // Assert
+
+            // Read response
+            byte[] buffer = new byte[1024];
+            var length = await responseContent.ReadAsync(buffer).AsTask().WithTimeout();
+            Assert.Equal("POST Response", Encoding.UTF8.GetString(buffer, 0, length));
+
+            // Ensure request stream has started
+            await requestStreamSyncPoint.WaitForSyncPoint();
+
+            // Send content and finish request body
+            await requestStream.WriteAsync(Encoding.UTF8.GetBytes("Hello world")).AsTask().WithTimeout();
```

Comment:
Scenario I expect to hang: here you write more data than will fit into the pipe buffer and hit backpressure. The request finishes without reading the request data. The server hangs waiting for the request body to finish.

When the app finishes without draining the request body, the server needs to cancel the body, not drain it.

Reply:
Ok. How do I test if the request pipe still has data in it? Something like this?

```cs
if (!_requestPipe.Reader.TryRead(out var result) || !result.IsCompleted)
{
    AbortRequest();
}
else
{
    await _requestPipe.Writer.CompleteAsync();
    await _requestPipe.Reader.CompleteAsync();
}
```

Also how do you cancel the request body? `_requestPipe.Writer.CancelPendingFlush()` doesn't appear to cancel the blocked `_requestPipe.Writer.AsStream().WriteAsync(...)`.

Reply:
I wouldn't do try read. Just drain the thing or cancel it then await the task.

Reply:
It should be canceled if there is blocked content in the unit test because of flow control to propagate the error to the unit test (mirrors what would happen if writing to request when it was completed)

I could make it always abort (aka cancel read and flush). Why is that better?